### PR TITLE
[WIP] Expose functionality as a re-usable library

### PR DESCRIPTION
**Pull request by jiegec**
This pr implements #268 by splitting code into two parts: liblsof(the library) and lsof(the cli). The pr is unfinished, but is half way done. It is a VERY BIG pull request and I try my best to maintain the lsof cli behavior. I am opening this pr to request for more comments. Progress:

- [x] Re-arrange directory structure, build liblsof using libtool.
- [x] Add initial library interface with Doxygen.
- [x] Move global and static variables to lsof context.
- [x] Rewrite LTbasic & LTszoff with liblsof.
- [x] Set visibility to hidden for internal functions of liblsof.
- [x] Migrate to Linux/Darwin/FreeBSD/NetBSD/OpenBSD/Solaris/OpenIndiana. Support for AIX/HP-UX/SCO/UnixWare will be broken, but I do not have those environments to fix.
- [x] Write some utilities with liblsof as an example, maybe re-implement a netstat?
- [x] Selection result reporting, implemented as an array of tagged union, e.g. selection by pid success, selection by command name failed etc.
- [x] Add optional fields to `struct lsof_file`, e.g. network address, TCP state.
- [ ] Allow liblsof to read info when running as non-root on NetBSD and Solaris by switching to /proc-based.
- [ ] Replace all exit calls in liblsof with proper error reporting.
- [ ] Migrate more options from cli to liblsof until cli only uses public interface of liblsof.
- [ ] Fix memory leaks.

I'd love to see more comments on the API interface (You can find some at #268).

### CODE REVIEW

File: lib/dialects/aix/dnode.c [hunk 1]
```diff
@@ -155,17 +155,17 @@ struct gnode *ga; /* local gnode address */
         switch (f.set.l_type & (F_RDLCK | F_WRLCK)) {
 
         case F_RDLCK:
-            return ((l) ? 'R' : 'r');
+            return ((l) ? LSOF_LOCK_READ_FULL : LSOF_LOCK_READ_PARTIAL);
```
Comment (BenBE) [inline lib/dialects/aix/dnode.c:158]: Style question: Any particular reason for those extra parens in the lines below?
```suggestion
            return (l) ? LSOF_LOCK_READ_FULL : LSOF_LOCK_READ_PARTIAL;
```

File: lib/dialects/darwin/Mksrc [hunk 2]
```diff
@@ -12,13 +12,26 @@
...
+
+D=src
+L="arg.c main.c misc.c node.c print.c ptti.c store.c usage.c util.c"
+
+mksrc
+
+D=src/dialects/darwin
+L="dprint.c"
+
+mksrc
```
Comment (BenBE) [inline lib/dialects/darwin/Mksrc:37]: Maybe add a newline at EOF?

File: lib/dialects/freebsd/Mksrc [hunk 3]
```diff
@@ -12,14 +12,21 @@
 #
 # $Id: Mksrc,v 1.5 2008/04/15 13:31:47 abe Exp $
 
+mksrc() {
```
Comment (BenBE) [inline lib/dialects/freebsd/Mksrc:15]: Didn't another similar change cause this file to be shabanged as `/bin/bash` on another dialect?

**Review by BenBE — COMMENTED** [review]
The overall changes seem to be consistent and easy to follow for review (didn't manage to have a full look at everything, but the first ~20% seem sensible).

Minor remarks exist re code style, as well as some questions. Will continue my review later.

Looking at the changes overall leaves the impression, that this in places simplifies the existing code and removes a bit of repetition.

One implementation detail I noticed: Besides not calling `exit` inside the library, additionally writing to `stdout` and `stderr` should be handled via the context too (e.g. put some callback in the context to receive logging/error related information/output (e.g. dialect `hpux` -> `kmem/dproc.c`: The calls to `fprintf` inside `get_kernel_access` should be handled).

**Comment by jiegec** [comment]
> One implementation detail I noticed: Besides not calling `exit` inside the library, additionally writing to `stdout` and `stderr` should be handled via the context too (e.g. put some callback in the context to receive logging/error related information/output (e.g. dialect `hpux` -> `kmem/dproc.c`: The calls to `fprintf` inside `get_kernel_access` should be handled).

Sure, the current proposed solution is to output via `ctx->err` instead of `stderr`. The replace is not done yet because there are so many occurrences. If the library user wants to collect the output, `fmemopen` can be used.

**Comment by masatake** [comment]
Incredible work!

**Comment by BenBE** [comment]
Great work so far and even though I didn't continue yet with a follow up review for the parts I didn't get to in my first review, I'm still somewhat following this quite closely.

That said, it's kinda hard to review this set of changes realistically due to the sheer amount of affected files and number of commits. One theme I remarked in my first review concerns the tiny clean-ups that are happening along the way. They are necessary, but a more systematic way to know if a commit is just cleanup or affects the project structure would be good; ideally the mere fixups should be on a different PR so they can be skipped more easily. As much as I like this PR I don't feel confident in its current state that a review is likely to find any fallout or oder (un)intended (hidden) changes apart from the most basic ones. Thus it's probably best to mark this PR as draft for now and reorder/rework the individual commits once all pieces for this change are in place.

As far as I noticed in the commits so far, there are essentially four phases the commits can be sorted into:
* Cleanup / preparation
* Documentation updates
* Build system adjustments
* Code/data structure changes

Given that the last phase (actual code/data structure changes) is the main aspect of this PR, I'd somehow suggest having at least the cleanup and documentation issues be separated into a separate branch. This allows to have the project benefit from these (small) changes even long before the whole library PR is finalized.

I'll try to start a second attempt at reviewing the changes accumulated so far soon. So in case there's sth I should wait for before starting just drop me a note.

**Comment by BenBE** [comment]
NB: @jiegec Quick question as I don't seem to see it in the changes, but kinda expected to: Will there be some configuration file(s) for `pkg-config` with the new library, or will using the lib externally somewhat boil down on fixed (include) paths for finding the necessary ingredients?

**Comment by jiegec** [comment]
> NB: @jiegec Quick question as I don't seem to see it in the changes, but kinda expected to: Will there be some configuration file(s) for `pkg-config` with the new library, or will using the lib externally somewhat boil down on fixed (include) paths for finding the necessary ingredients?

Okay, I will add `.pc` file to locate the new library. I am currently working on the second part at branch library-part2: adding context arguments.

File: lib/dialects/hpux/kmem/dnode2.c [hunk 4]
```diff
@@ -73,8 +73,8 @@ typedef struct afs_lock afs_rwlock_t;
  * Local function prototypes
  */
 
-_PROTOTYPE(static struct volume *getvolume, (struct VenusFid * f, int *vols));
-_PROTOTYPE(static int is_rootFid, (struct vcache * vc, int *rfid));
+static struct volume *getvolume, (struct VenusFid * f int *vols);
+static int is_rootFid, (struct vcache * vc int *rfid);
```
Comment (BenBE) [inline lib/dialects/hpux/kmem/dnode2.c:77]: Stray comma

File: lib/dialects/freebsd/dsock.c [hunk 5]
```diff
@@ -82,8 +82,7 @@ static char copyright[] =
  * Local function prototypes
  */
 
-_PROTOTYPE(static int ckstate,
-           (struct lsof_context * ctx, struct xtcpcb *pcb, int fam));
+static int ckstate           (struct lsof_context * ctx, struct xtcpcb *pcb, int fam);
```
Comment (BenBE) [inline lib/dialects/freebsd/dsock.c]: Remove whitespace?

File: lib/dialects/freebsd/dproto.h [hunk 6]
```diff
@@ -35,45 +35,37 @@
  */
 
 #if !defined(N_UNIX)
-_PROTOTYPE(extern char *get_nlist_path, (struct lsof_context * ctx, int ap));
+extern char *get_nlist_path, (struct lsof_context * ctx int ap);
```
Comment (BenBE) [inline lib/dialects/freebsd/dproto.h]: Stray comma

File: lib/dialects/aix/dproto.h [hunk 7]
```diff
@@ -38,30 +38,28 @@
...
-_PROTOTYPE(extern int readafsnode,
-           (KA_T va, struct vnode *v, struct afsnode *an));
+extern struct vnode *alloc_vcache(void);
+extern int hasAFS(struct vnode *vp);
+extern int readafsnode(KA_T va, struct vnode *v, struct afsnode *an);
 #    endif /* defined(HAS_AFS) */
 
 #    if defined(HAS_JFS2)
-_PROTOTYPE(extern int readj2lino, (struct gnode * ga, struct l_ino *li));
+extern int readj2lino, (struct gnode * ga struct l_ino * li);
```
Comment (BenBE) [inline lib/dialects/aix/dproto.h]: Stray comma

File: lib/dialects/sun/dnode.c [hunk 8]
```diff
@@ -411,95 +410,88 @@ static v_optab_t **Voptab = (v_optab_t **)NULL;
...
+                    struct pid *pids);
 #endif /* defined(HASPROCFS) */
 
-_PROTOTYPE(static char *ent_fa, (KA_T * a1, KA_T *a2, char *d, int *len));
-_PROTOTYPE(static int is_socket, (struct lsof_context * ctx, struct vnode *v));
-_PROTOTYPE(static int read_cni,
-           (struct lsof_context * ctx, struct snode *s, struct vnode *rv,
-            struct vnode *v, struct snode *rs, struct dev_info *di, char *din,
-            int dinl));
+static char *ent_fa, (KA_T * a1, KA_T *a2, char *d int *len);
```
Comment (BenBE) [inline lib/dialects/sun/dnode.c:475]: Stray comma

File: lib/dialects/sun/dnode1.c [hunk 9]
```diff
@@ -112,8 +112,8 @@ typedef struct afs_bozoLock afs_bozoLock_t;
  * Local function prototypes
  */
 
-_PROTOTYPE(static struct volume *getvolume, (struct VenusFid * f, int *vols));
-_PROTOTYPE(static int is_rootFid, (struct vcache * vc, int *rfid));
+static struct volume *getvolume, (struct VenusFid * f int *vols);
+static int is_rootFid, (struct vcache * vc int *rfid);
```
Comment (BenBE) [inline lib/dialects/sun/dnode1.c:116]: Stray comma

File: lib/dialects/sun/dnode2.c [hunk 10]
```diff
@@ -44,9 +44,8 @@ static char copyright[] =
 #        include <vxfsutil.h>
 #        define EMSGPFX "vx_inode: "
 
-_PROTOTYPE(static char *add2em, (char *em, char *fmt, char *arg));
-_PROTOTYPE(static char *ckptr,
-           (char *em, char *ptr, int len, int slen, char *nm));
+static char *add2em, (char *em, char *fmt char *arg);
```
Comment (BenBE) [inline lib/dialects/sun/dnode2.c:47]: Stray comma

File: lib/dialects/sun/dnode2.c [hunk 11]
```diff
@@ -44,9 +44,8 @@ static char copyright[] =
...
+static char *ckptr           (char *em, char *ptr, int len, int slen, char *nm);
```
Comment (BenBE) [inline lib/dialects/sun/dnode2.c]: Remove whitespace?

File: lib/dialects/sun/dproto.h [hunk 12]
```diff
@@ -35,78 +35,73 @@
...
 
-_PROTOTYPE(extern int is_file_named,
-           (struct lsof_context * ctx, char *p, int nt, enum vtype vt, int ps));
-_PROTOTYPE(extern struct l_vfs *readvfs, (struct lsof_context * ctx, KA_T ka,
-                                          struct vfs *la, struct vnode *lv));
-_PROTOTYPE(extern int vop2ty,
-           (struct lsof_context * ctx, struct vnode *vp, int fx));
+extern int is_file_named(struct lsof_context *ctx, char *p, int nt,
+                         enum vtype vt, int ps);
+extern struct l_vfs *readvfs,
```
Comment (BenBE) [inline lib/dialects/sun/dproto.h:54]: Stray comma

File: lib/dialects/sun/dproto.h [hunk 13]
```diff
@@ -35,78 +35,73 @@
...
-_PROTOTYPE(extern void clr_sect, (void));
-_PROTOTYPE(extern int rw_pseudo_sect, (struct lsof_context * ctx, int m));
+extern int rw_clone_sect(struct lsof_context *ctx, int m);
+extern void clr_sect(void);
+extern int rw_pseudo_sect(struct lsof_context *ctx, int m);
 #endif /* defined(HASDCACHE) */
 
 #if defined(HASIPv6)
-_PROTOTYPE(extern struct hostent *gethostbyname2, (const char *nm, int proto));
+extern struct hostent *gethostbyname2, (const char *nm int proto);
```
Comment (BenBE) [inline lib/dialects/sun/dproto.h:72]: Stray comma

File: lib/dialects/sun/dproto.h [hunk 14]
```diff
@@ -35,78 +35,73 @@
...
-_PROTOTYPE(extern void process_socket,
-           (struct lsof_context * ctx, KA_T sa, char *ty));
+extern void close_kvm(struct lsof_context *ctx);
+extern void open_kvm(struct lsof_context *ctx);
+extern void process_socket(struct lsof_context *ctx, KA_T sa, char *ty);
 
 #if solaris >= 110000
-_PROTOTYPE(extern int process_VSOCK, (struct lsof_context * ctx, KA_T va,
-                                      struct vnode *v, struct sonode *so));
+extern int process_VSOCK,
```
Comment (BenBE) [inline lib/dialects/sun/dproto.h:94]: Stray comma

File: lib/dialects/sun/dsock.c [hunk 15]
```diff
@@ -227,12 +227,9 @@ static CTF_request_t IRU_requests[] = {{ICMP_T_TYPE_NAME, icmp_t_members},
...
 
-_PROTOTYPE(static int read_icmp_t,
-           (struct lsof_context * ctx, KA_T va, KA_T ph, KA_T ia, icmp_t *ic));
-_PROTOTYPE(static int read_rts_t,
-           (struct lsof_context * ctx, KA_T va, KA_T ph, KA_T ra, rts_t *rt));
-_PROTOTYPE(static int read_udp_t,
-           (struct lsof_context * ctx, KA_T ua, udp_t *uc));
+static int read_icmp_t           (struct lsof_context * ctx, KA_T va, KA_T ph, KA_T ia, icmp_t *ic);
+static int read_rts_t           (struct lsof_context * ctx, KA_T va, KA_T ph, KA_T ra, rts_t *rt);
+static int read_udp_t           (struct lsof_context * ctx, KA_T ua, udp_t *uc);
```
Comment (BenBE) [inline lib/dialects/sun/dsock.c]: Remove whitespace?

File: lib/proto.h [hunk 16]
```diff
@@ -62,102 +49,95 @@
...
-_PROTOTYPE(extern void clear_pinfo, (struct lsof_context * ctx));
-_PROTOTYPE(extern pxinfo_t *find_pepti, (struct lsof_context * ctx, int pid,
-                                         struct lfile *lf, pxinfo_t *pp));
-_PROTOTYPE(extern void process_pinfo, (struct lsof_context * ctx, int f));
-_PROTOTYPE(extern void clear_psxmqinfo, (struct lsof_context * ctx));
-_PROTOTYPE(extern pxinfo_t *find_psxmqinfo, (struct lsof_context * ctx, int pid,
-                                             struct lfile *lf, pxinfo_t *pp));
-_PROTOTYPE(extern void process_psxmqinfo, (struct lsof_context * ctx, int f));
+extern void clear_pinfo(struct lsof_context *ctx);
+extern pxinfo_t *find_pepti,
```
Comment (BenBE) [inline lib/proto.h]: Stray comma

File: tests/LTdnlc.c [hunk 17]
```diff
@@ -81,8 +81,8 @@ char *Pn = (char *)NULL; /* program name */
  * Local function prototypes
  */
 
-_PROTOTYPE(static void cleanup, (void));
-_PROTOTYPE(static char *FindLsofCwd, (int *ff, LTdev_t *cwddc, char *ibuf));
+static void cleanup (void);
+static char *FindLsofCwd, (int *ff, LTdev_t *cwddc char *ibuf);
```
Comment (BenBE) [inline tests/LTdnlc.c]: Stray comma

File: tests/LTlib.c [hunk 18]
```diff
@@ -199,11 +199,10 @@ static int Ufo = 0;                     /* Fo[] structures used */
...
 
-_PROTOTYPE(static void closepipe, (void));
-_PROTOTYPE(static void getlsofpath, (void));
-_PROTOTYPE(static int GetOpt,
-           (int ct, char *opt[], char *rules, char **em, char *pn));
-_PROTOTYPE(static X2DEV_T x2dev, (char *x, char **em));
+static void closepipe (void);
+static void getlsofpath (void);
+static int GetOpt           (int ct, char *opt[], char *rules, char **em, char *pn);
+static X2DEV_T x2dev, (char *x char **em);
```
Comment (BenBE) [inline tests/LTlib.c]: Stray comma

File: tests/LTlib.c [hunk 19]
```diff
@@ -199,11 +199,10 @@ static int Ufo = 0;                     /* Fo[] structures used */
...
  */
 
-_PROTOTYPE(static void closepipe, (void));
-_PROTOTYPE(static void getlsofpath, (void));
-_PROTOTYPE(static int GetOpt,
-           (int ct, char *opt[], char *rules, char **em, char *pn));
-_PROTOTYPE(static X2DEV_T x2dev, (char *x, char **em));
+static void closepipe (void);
+static void getlsofpath (void);
+static int GetOpt           (int ct, char *opt[], char *rules, char **em, char *pn);
```
Comment (BenBE) [inline tests/LTlib.c]: Remove whitespace?

File: tests/LTlock.c [hunk 20]
```diff
@@ -206,10 +206,10 @@ char *Pn = (char *)NULL;   /* program name */
  * Local function prototypes
  */
 
-_PROTOTYPE(static void cleanup, (void));
-_PROTOTYPE(static char *lkfile, (int ty));
-_PROTOTYPE(static char *tstwlsof, (char *opt, char *xlk));
-_PROTOTYPE(static char *unlkfile, (int ty));
+static void cleanup (void);
+static char *lkfile (int ty);
+static char *tstwlsof, (char *opt char *xlk);
```
Comment (BenBE) [inline tests/LTlock.c]: Stray comma

File: tests/LTnlink.c [hunk 21]
```diff
@@ -68,9 +68,8 @@ char *Pn = (char *)NULL;   /* program name */
  * Local function prototypes
  */
 
-_PROTOTYPE(static void cleanup, (void));
-_PROTOTYPE(static char *FindFile, (char *opt, int *ff, int ie, LTdev_t *tfdc,
-                                   char *ibuf, char *xlnk, char *szbuf));
+static void cleanup (void);
+static char *FindFile, (char *opt, int *ff, int ie, LTdev_t *tfdc                                   char *ibuf, char *xlnk, char *szbuf);
```
Comment (BenBE) [inline tests/LTnlink.c]: Stray comma

File: tests/LTszoff.c [hunk 22]
```diff
@@ -76,8 +76,8 @@ char *Pn = (char *)NULL;   /* program name */
  * Local function prototypes
  */
 
-_PROTOTYPE(static void cleanup, (void));
-_PROTOTYPE(static char *testlsof, (int tt, char *opt, char *xval));
+static void cleanup (void);
+static char *testlsof, (int tt, char *opt char *xval);
```
Comment (BenBE) [inline tests/LTszoff.c]: Stray comma

File: tests/LsofTest.h [hunk 23]
```diff
@@ -330,17 +330,17 @@ extern int errno; /* error number */
...
+extern char *ConvLsofDev, (char *dev LTdev_t *ldev);
+extern char *ExecLsof (char **opt);
+extern char *IsLsofExec (void);
+extern void LTlibClean (void);
+extern char *MkStrCpy, (char *src int *len);
+extern LTfldo_t *RdFrLsof, (int *nf char **em);
+extern void PrtMsg, (char *mp char *pn);
+extern void PrtMsgX, (char *mp, char *pn, void (*f)() int xv);
+extern int ScanArg, (int ac, char *av[], char *opt char *pn);
+extern void StopLsof (void);
```
Comment (BenBE) [inline tests/LsofTest.h]: Stray comma

**Review by BenBE — CHANGES_REQUESTED** [review]
Please be more careful with your changes.

File: lib/dialects/hpux/kmem/dnode1.c [hunk 24]
```diff
@@ -98,16 +98,17 @@ typedef int time_t;
...
-                int *devs,         /* device status receiver */
-                dev_t *rdev,       /* raw device number receiver */
-                int *rdevs)        /* raw device status receiver */
+int read_vxnode(v, vfs, dev, devs, rdev, rdevs)
+struct vnode *v;   /* local containing vnode */
+struct l_vfs *vfs; /* local vfs structure */
+dev_t *dev;        /* device number receiver */
+int *devs;         /* device status receiver */
+dev_t *rdev;       /* raw device number receiver */
+int *rdevs;        /* raw device status receiver */
```
Comment (BenBE) [inline lib/dialects/hpux/kmem/dnode1.c:107]: Why not rewrite this to more modern C like with so many other function declarations?

**Comment by jiegec** [comment]
> Please be more careful with your changes.

Thanks, I was removing `_PROTOTYPE` via regex, but the matching does not work well on multiline programs.. I do not plan to merge this large pr, but take effort to complete in parts.

**Review by BenBE — COMMENTED** [review]
Why are some function signatures reverted back to K&R style?

**Comment by BenBE** [comment]
NB: Ah, k. Maybe have a look at `git rerere`, which should help with such situations. Also I noticed `kdiff3` usually does a quite good job at resolving most conflicts correctly automatically.

**Comment by BenBE** [comment]
@jiegec Any further open aspects/items for this feature? AFAICS the bullet list in the first post has not been checked off completely.

**Pull request closed without merging**